pyomo/contrib/preprocessing/plugins/induced_linearity.py

```diff
@@ -0,0 +1,325 @@
+            ({v2lb}, {v2ub}).
+            """.format(v1=v1, v2=v2, v2lb=v2.lb, v2ub=v2.ub)).strip())
+        return False
+    blk = Block()
+    unique_name = unique_component_name(
+        block, ("%s_%s_bilinear" % (v1.local_name, v2.local_name))
+        .replace('[', '').replace(']', ''))
+    block._induced_linearity_info.add_component(unique_name, blk)
+    # TODO think about not using floats as indices in a set
+    blk.valid_values = Set(initialize=var_values)
+    blk.x_active = Var(blk.valid_values, domain=Binary, initialize=1)
+    blk.v_increment = Var(
+        blk.valid_values, domain=v2.domain,
+        bounds=(v2.lb, v2.ub), initialize=v2.value)
+    blk.v_defn = Constraint(expr=v2 == summation(blk.v_increment))
+
+    @blk.Constraint(blk.valid_values)
+    def v_lb(blk, val):
+        return v2.lb * blk.x_active[val] <= blk.v_increment[val]
+
+    @blk.Constraint(blk.valid_values)
+    def v_ub(blk, val):
+        return blk.v_increment[val] <= v2.ub * blk.x_active[val]
+    blk.select_one_value = Constraint(expr=summation(blk.x_active) == 1)
+    # Categorize as case 1 or case 2
+    for bilinear_constr in bilinear_constrs:
+        # repn = generate_standard_repn(bilinear_constr.body)
+
+        # Case 1: no other variables besides bilinear term in constraint. v1
+        # (effectively discrete variable) is positive.
+        # if (len(repn.quadratic_vars) == 1 and len(repn.linear_vars) == 0
+        #         and repn.nonlinear_expr is None):
+        #     _reformulate_case_1(v1, v2, discrete_constr, bilinear_constr)
+
+        # NOTE: Case 1 is left unimplemented for now, because it involves some
+        # messier logic with respect to how the transformation needs to happen.
+
+        # Case 2: this is everything else, but do we want to have a special
+        # case if there are nonlinear expressions involved with the constraint?
+        pass
+        _reformulate_case_2(blk, v1, v2, bilinear_constr)
+    pass
+
+
+def _reformulate_case_2(blk, v1, v2, bilinear_constr):
+    repn = generate_standard_repn(bilinear_constr.body)
+    replace_index = next(
+        i for i, var_tup in enumerate(repn.quadratic_vars)
+        if (var_tup[0] is v1 and var_tup[1] is v2) or
+           (var_tup[0] is v2 and var_tup[1] is v1))
+    bilinear_constr.set_value((
+        bilinear_constr.lower,
+        sum(coef * repn.linear_vars[i]
+            for i, coef in enumerate(repn.linear_coefs)) +
+        repn.quadratic_coefs[replace_index] * sum(
+            val * blk.v_increment[val] for val in blk.valid_values) +
+        sum(repn.quadratic_coefs[i] * var_tup[0] * var_tup[1]
+            for i, var_tup in enumerate(repn.quadratic_vars)
+            if not i == replace_index) +
+        repn.constant +
+        zero_if_None(repn.nonlinear_expr),
+        bilinear_constr.upper
+    ))
+
+
+def zero_if_None(val):
+    return val if val is not None else 0
```

Comment:
Any particular reason the logic here is "reversed" from the function name?  I probably would have implemented it `return 0 if val is None else val`.